### PR TITLE
Unit test framework

### PR DESCRIPTION
**Pull request by junyk**
Unit test framework:
* Basic definitions test
* A runner of test suite of individual tools
* Logic on how to apply test cases and check their expected output

### CODE REVIEW

File: janis_core/tool/test_classes.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    def __init__(
+        self,
+        tag: str,
+        compared: Enum,
+        operator: Callable,
+        expected_value: Optional[Any] = None,
+        expected_file: Optional[str] = None,
+        file_diff_source: Optional[str] = None,
+        array_index: Optional[int] = None,
+        suffix: Optional[str] = None,
```
Comment (junyk) [inline janis_core/tool/test_classes.py:32]: call this suffix_secondary_file

File: janis_core/tool/test_classes.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ):
+        self.tag = tag
+        self.compared = compared
+        self.operator = operator
+        self.expected_value = expected_value
+
+        # we can have a 'string' expected output stored in a file
+        self.expected_file = expected_file
```
Comment (junyk) [inline janis_core/tool/test_classes.py:40]: Add a check for either one of expected value or expected file

File: janis_core/tool/test_suite_runner.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        test_helpers.verify_janis_assistant_installed()
+
+        from janis_assistant.main import run_with_outputs
+        from janis_assistant.management.configuration import JanisConfiguration
+
+        output_dir = os.path.join(os.getcwd(), "tests_output", self.tool.id())
+        config = JanisConfiguration(engine=engine)
+        output = run_with_outputs(
+            tool=self.tool, inputs=input, output_dir=output_dir, config=config
```
Comment (junyk) [inline janis_core/tool/test_suite_runner.py:37]: Michael will add an `engine` input parameter here, so i don't need to set the JanisConfiguration.

File: janis_core/tool/test_helpers.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+def get_available_engines():
+    """
+    Get a list of available engines to run the test suite against
+    """
+    verify_janis_assistant_installed()
+    from janis_assistant.engines.enginetypes import EngineType
+
+    engines = {
+        EngineType.cromwell.value: WdlTranslator(),
+        EngineType.cwltool.value: CwlTranslator(),
```
Comment (junyk) [inline janis_core/tool/test_helpers.py:43]: There is a list of translations called supported translations in janis_core

File: janis_core/tool/test_helpers.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    except Exception as e:
+        raise e
+
+
+def get_available_engines():
+    """
+    Get a list of available engines to run the test suite against
+    """
+    verify_janis_assistant_installed()
+    from janis_assistant.engines.enginetypes import EngineType
```
Comment (junyk) [inline janis_core/tool/test_helpers.py:39]: There is an `EngineType.all()`

File: janis_core/tool/test_definitions.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        # Call this outside the try-except so that we can still throw
+        # different exceptions relevant to the actual logic of this function
+        if not tool.container():
+            print(tool.id(), "no container")
+        else:
+            print(tool.id(), tool.container())
+
+        ci = ContainerInfo.parse(tool.container())
+        registry = ContainerRegistry.from_host(ci.host).to_registry()
+        digest = registry.get_digest(ci)
```
Comment (junyk) [inline janis_core/tool/test_definitions.py:146]: https://github.com/PMCC-BioinformaticsCore/janis-assistant/blob/janis-prepare/janis_assistant/data/container/__init__.py#L29

Comment (junyk) [reply janis_core/tool/test_definitions.py:146]: use the cache directory

File: janis_core/tool/test_definitions.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        output_dir = os.path.join(os.getcwd(), "tests_output", tool.id())
+
+        errors = []
+        for engine in engines:
+            try:
+                translator = engines[engine]
+                translator.translate(
+                    tool,
+                    export_path=output_dir,
+                    should_validate=True,
```
Comment (junyk) [inline janis_core/tool/test_definitions.py:168]: Skip validation for now. We will revisit.

File: janis_core/tool/test_classes.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+from enum import Enum
+from typing import Optional, Dict, Any, Callable
+
+
+class TTestCompared(Enum):
```
Comment (junyk) [inline janis_core/tool/test_classes.py:5]: rename to `TTestValuePreProcessor`

File: janis_core/tool/test_classes.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    Describe the logic on how to test the expected output of a test case.
+    A test case can have multiple instances of this class to test different output or
+    different logic of the same output
+    """
+
+    def __init__(
+        self,
+        tag: str,
+        compared: Enum,
```
Comment (junyk) [inline janis_core/tool/test_classes.py:26]: pre_processor

File: janis_core/tool/test_classes.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        operator: Callable,
```
Comment (junyk) [inline janis_core/tool/test_classes.py:27]: rename `comparator`

File: janis_core/tool/test_suite_runner.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            test_logic=test_logic,
+            output_value=output_value,
+            output_type=output_type,
+        )
+
+        return self._transform_value(
+            test_logic=test_logic,
+            output_value=output_value,
+            output_type=output_type,
+        )
```
Comment (junyk) [inline janis_core/tool/test_suite_runner.py:100]: We will choose to call `self._transform_value` or the callable passed to this object

**Review by illusional — APPROVED** [review]

**Pull request merged**